File: aif/tokens.py
"""Per-agent tokens: derivation, issuance, lookup and cascade revocation.

The gatekeeper is a *config* token (``AIF_ADMIN_TOKEN``) and has no row here; everything it issues
becomes a tree root. Every claimed token holder may issue children under its own token, and any
ancestor (or the gatekeeper) may revoke an entire subtree - a parent answers for everything below it.

Token shape: ``aif_`` + 24 hex chars of ``sha256(salt \\0 name \\0 nonce)``. Tokens are stored
cleartext (they grant what they grant), never logged, and only ever *returned* at issue/claim time.
An invite carries no name: the holder registers a name with it, and the server then rewrites the
row to the final name-derived token, which is what the agent uses from then on.
"""

from __future__ import annotations

import hashlib
import secrets
import sqlite3
from typing import Any

from . import db
from .config import Config

TOKEN_LEN = 24


def derive_token(salt: str, name: str, nonce: str = "") -> str:
    """Deterministic token for a (name, nonce) pair under this server's salt."""
    digest = hashlib.sha256(f"{salt}\0{name}\0{nonce}".encode()).hexdigest()
    return f"aif_{digest[:TOKEN_LEN]}"


def new_nonce() -> str:
    return secrets.token_hex(8)


def lookup(conn: sqlite3.Connection, token: str) -> dict[str, Any] | None:
    row = conn.execute("SELECT * FROM tokens WHERE self_token = ?", [token]).fetchone()
    return dict(row) if row else None


def issue(cfg: Config, conn: sqlite3.Connection, issuer: dict[str, Any] | None, name: str = "", descr: str = "", days: float | None = None) -> dict[str, Any]:
    """Create a token row. ``issuer=None`` means the gatekeeper (a root); otherwise the issuer's
    row, so the child hangs under it. Named tokens skip the name-choice step but still get claimed
    on first use."""
    nonce = new_nonce()
    token = derive_token(cfg.token_salt, name, nonce)
    now = db.now()
    if issuer is None:  # a tree root: parent == root == self
        root = parent = token
    else:
        parent, root = issuer["self_token"], issuer["root_token"]
    exp = now + days * 86400 if days else (0.0 if name else now + cfg.invite_ttl)
    conn.execute(
        "INSERT INTO tokens (name, low, root_token, parent_token, self_token, descr, created, exp, nonce) VALUES (?,?,?,?,?,?,?,?,?)",
        [name, name.lower(), root, parent, token, descr, now, exp, nonce],
    )
    return {"token": token, "name": name, "parent": parent, "root": root, "exp": exp, "nonce": nonce}


def claim(cfg: Config, conn: sqlite3.Connection, row: dict[str, Any], name: str) -> str:
    """Bind *name* to a token row and return the final, name-derived token.

    The invite form of the token stops existing here: self_token is rewritten to
    ``derive(salt, name, nonce)`` (root tokens keep root == parent == self). No other row can
    reference the old value: children require a claimed parent, so none exist yet.
    """
    final = derive_token(cfg.token_salt, name, row["nonce"])
    if row["root_token"] == row["self_token"]:  # a root: its own root/parent pointers move too
        conn.execute(
            "UPDATE tokens SET name = ?, low = ?, self_token = ?, claimed = ?, root_token = ?, parent_token = ? WHERE self_token = ?",
            [name, name.lower(), final, db.now(), final, final, row["self_token"]],
        )
    else:
        conn.execute(
            "UPDATE tokens SET name = ?, low = ?, self_token = ?, claimed = ? WHERE self_token = ?",
            [name, name.lower(), final, db.now(), row["self_token"]],
        )
    return final


def check_live(conn: sqlite3.Connection, row: dict[str, Any] | None, code_unknown: str = "bad_token") -> dict[str, Any]:
    """Raise the precise reason a token row is unusable, or hand it back."""
    from .core import ApiError  # local import: core imports this module's helpers

    if row is None:
        raise ApiError(403, code_unknown, "access token rejected", "use the server's AIF_TOKEN value, or an issued agent token")
    if row["revoked"]:
        raise ApiError(403, "token_revoked", "this token was revoked", "ask your issuer (or the gatekeeper) for a fresh one")
    if row["exp"] and row["exp"] < db.now():
        if row["claimed"]:
            raise ApiError(403, "token_expired", "this token expired", "ask your issuer (or the gatekeeper) for a fresh one")
        raise ApiError(403, "invite_expired", "this invite was never claimed and expired", "ask your issuer for a fresh invite")
    return row


def children_of(conn: sqlite3.Connection, self_token: str) -> list[dict[str, Any]]:
    return [dict(r) for r in conn.execute("SELECT * FROM tokens WHERE parent_token = ? AND self_token != parent_token", [self_token])]


def subtree(conn: sqlite3.Connection, self_token: str) -> list[dict[str, Any]]:
    """The node itself plus every descendant, following parent_token downward."""
    me = lookup(conn, self_token)
    out: dict[str, dict[str, Any]] = {self_token: me} if me else {}
    frontier = [self_token]
    while frontier:
        batch = frontier
        frontier = []
        for row in conn.execute(f"SELECT * FROM tokens WHERE parent_token IN ({db.marks(batch)})", batch):
            if row["self_token"] in out:
                continue
            out[row["self_token"]] = dict(row)
            frontier.append(row["self_token"])
    return list(out.values())


def is_ancestor(conn: sqlite3.Connection, ancestor_token: str, target: dict[str, Any]) -> bool:
    """Walk *target*'s parent chain to its root; True if *ancestor_token* shows up on the way."""
    seen: set[str] = set()
    current = target
    while True:
        if current["self_token"] == ancestor_token:
            return True
        if current["parent_token"] in (current["self_token"],) or current["self_token"] in seen:
            return False
        seen.add(current["self_token"])
        parent = lookup(conn, current["parent_token"])
        if parent is None:
            return False
        current = parent


def revoke_subtree(conn: sqlite3.Connection, target: dict[str, Any]) -> list[str]:
    """Cascade: mark the target and its whole subtree revoked; returns the affected token rows' ids."""
    nodes = subtree(conn, target["self_token"])
    now = db.now()
    for node in nodes:
        if not node["revoked"]:
            conn.execute("UPDATE tokens SET revoked = ? WHERE self_token = ?", [now, node["self_token"]])
    return [n["self_token"] for n in nodes]

File: internal/tokens/tokens.go
package tokens

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"

	"github.com/dshein-alt/aif/internal/config"
	"github.com/dshein-alt/aif/internal/db"
)

const TokenLen = 24

// LiveSQL is the one definition of "live" for a token row: not revoked and not expired.
const LiveSQL = "revoked IS NULL AND (exp = 0 OR exp > ?)"

func IsLive(row map[string]any, ts float64) bool {
	if row == nil {
		return false
	}
	if !db.IsNull(row, "revoked") && db.AsFloat(row, "revoked") != 0 {
		return false
	}
	exp := db.AsFloat(row, "exp")
	return exp == 0 || exp > ts
}

// DeriveToken returns the deterministic token for a (name, nonce) pair under this server's salt.
func DeriveToken(salt, name, nonce string) string {
	sum := sha256.Sum256(fmt.Appendf(nil, "%s\x00%s\x00%s", salt, name, nonce))
	return "aif_" + hex.EncodeToString(sum[:])[:TokenLen]
}

func NewNonce() string {
	b := make([]byte, 8)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func Lookup(ctx context.Context, d db.DB, token string) (map[string]any, error) {
	return db.QueryOne(ctx, d, "SELECT * FROM tokens WHERE self_token = ?", token)
}

type Issued struct {
	Token  string
	Name   string
	Parent string
	Root   string
	Exp    float64
	Nonce  string
}

// Issue creates a token row. issuer==nil means a tree root (parent==root==self).
func Issue(ctx context.Context, d db.DB, cfg *config.Config, issuer map[string]any, name, descr string, days float64) (*Issued, error) {
	nonce := NewNonce()
	token := DeriveToken(cfg.TokenSalt, name, nonce)
	now := db.Now()
	root, parent := token, token
	if issuer != nil {
		parent = db.AsString(issuer, "self_token")
		root = db.AsString(issuer, "root_token")
	}
	var exp float64
	if days != 0 {
		exp = now + days*86400
	} else if name == "" {
		exp = now + float64(cfg.InviteTTL)
	}
	_, err := db.Exec(ctx, d,
		`INSERT INTO tokens (name, low, root_token, parent_token, self_token, descr, created, exp, nonce)
		 VALUES (?,?,?,?,?,?,?,?,?)`,
		name, name, root, parent, token, descr, now, exp, nonce)
	if err != nil {
		return nil, err
	}
	return &Issued{Token: token, Name: name, Parent: parent, Root: root, Exp: exp, Nonce: nonce}, nil
}

// Claim binds name to a token row and rewrites self_token to the name-derived form; returns it.
func Claim(ctx context.Context, d db.DB, cfg *config.Config, row map[string]any, name string) (string, error) {
	final := DeriveToken(cfg.TokenSalt, name, db.AsString(row, "nonce"))
	now := db.Now()
	if db.AsString(row, "root_token") == db.AsString(row, "self_token") {
		if _, err := db.Exec(ctx, d,
			`UPDATE tokens SET name=?, low=?, self_token=?, claimed=?, root_token=?, parent_token=? WHERE self_token=?`,
			name, name, final, now, final, final, db.AsString(row, "self_token")); err != nil {
			return "", err
		}
	} else {
		if _, err := db.Exec(ctx, d,
			`UPDATE tokens SET name=?, low=?, self_token=?, claimed=? WHERE self_token=?`,
			name, name, final, now, db.AsString(row, "self_token")); err != nil {
			return "", err
		}
	}
	// An un-named invite never expires once claimed (the invite TTL was a claim window, not a life).
	if db.AsString(row, "name") == "" {
		if _, err := db.Exec(ctx, d, "UPDATE tokens SET exp = 0 WHERE self_token = ?", final); err != nil {
			return "", err
		}
	}
	return final, nil
}

// Subtree returns the node plus every descendant, following parent_token downward.
func Subtree(ctx context.Context, d db.DB, selfToken string) ([]map[string]any, error) {
	me, err := Lookup(ctx, d, selfToken)
	if err != nil {
		return nil, err
	}
	out := map[string]map[string]any{}
	if me != nil {
		out[selfToken] = me
	}
	frontier := []string{selfToken}
	for len(frontier) > 0 {
		batch := frontier
		frontier = nil
		rows, err := db.QueryRows(ctx, d,
			fmt.Sprintf("SELECT * FROM tokens WHERE parent_token IN (%s)", db.Marks(len(batch))),
			toArgs(batch)...)
		if err != nil {
			return nil, err
		}
		for _, r := range rows {
			st := db.AsString(r, "self_token")
			if _, ok := out[st]; ok {
				continue
			}
			out[st] = r
			frontier = append(frontier, st)
		}
	}
	var res []map[string]any
	for _, v := range out {
		res = append(res, v)
	}
	return res, nil
}

// IsAncestor walks target's parent chain to its root; true if ancestorToken appears on the way.
func IsAncestor(ctx context.Context, d db.DB, ancestorToken string, target map[string]any) (bool, error) {
	seen := map[string]bool{}
	current := target
	for {
		if db.AsString(current, "self_token") == ancestorToken {
			return true, nil
		}
		if db.AsString(current, "parent_token") == db.AsString(current, "self_token") || seen[db.AsString(current, "self_token")] {
			return false, nil
		}
		seen[db.AsString(current, "self_token")] = true
		parent, err := Lookup(ctx, d, db.AsString(current, "parent_token"))
		if err != nil {
			return false, err
		}
		if parent == nil {
			return false, nil
		}
		current = parent
	}
}

// RevokeSubtree marks the target and its whole subtree revoked; returns affected self_tokens.
func RevokeSubtree(ctx context.Context, d db.DB, target map[string]any) ([]string, error) {
	nodes, err := Subtree(ctx, d, db.AsString(target, "self_token"))
	if err != nil {
		return nil, err
	}
	now := db.Now()
	affected := make([]string, 0, len(nodes))
	for _, n := range nodes {
		if db.IsNull(n, "revoked") {
			if _, err := db.Exec(ctx, d, "UPDATE tokens SET revoked = ? WHERE self_token = ?", now, db.AsString(n, "self_token")); err != nil {
				return nil, err
			}
		}
		affected = append(affected, db.AsString(n, "self_token"))
	}
	return affected, nil
}

func toArgs(s []string) []any {
	out := make([]any, len(s))
	for i, v := range s {
		out[i] = v
	}
	return out
}
